frontend_server/server.js

```diff
@@ -32,6 +34,31 @@ const HOST = '0.0.0.0';
 
 const app = express();
 
+// Add Authorization header for all requests that are proxied to the data server.
+// TODO: The token can be cached and only refreshed when needed
+app.use((req, res, next) => {
+  if (req.path.startsWith('/api') && assertEnvVar("NODE_ENV") === 'production') {
+    // Set up metadata server request
+    // See https://cloud.google.com/compute/docs/instances/verifying-instance-identity#request_signature
+    const metadataServerTokenURL = 'http://metadata/computeMetadata/v1/instance/service-accounts/default/identity?audience=';
+    const targetUrl = assertEnvVar("DATA_SERVER_URL");
+    const fetchUrl = metadataServerTokenURL + targetUrl;
+    const options = {
+      headers: {
+        'Metadata-Flavor': 'Google'
+      }
+    };
+    fetch(fetchUrl, options)
+      .then(res => res.text())
+      .then(token => {
+        req.headers["Authorization"] = `bearer ${token}`;
+        next();
+    });
```

Comment:
Add `.catch(next)` at the end of the promise chain to ensure asynchronous errors get handled correctly (see https://expressjs.com/en/guide/error-handling.html).